vignettes/tech-dm-keyed.Rmd

```diff
@@ -0,0 +1,216 @@
+---
+title: "Manipulating individual tables"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Technical: Manipulating individual tables}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+``````{r setup, include = FALSE}
+source("setup/setup.R")
+``````
+
+This vignette deals with situations where you want to transform tables of your `dm` and then update an existing table or add a new table to the `dm`.
+There are two straightforward approaches:
+
+1. individually subsetting the tables relevant to the calculation, perform the necessary transformations, add the result to the `dm` (or replace an existing table), and establish the key relations.
+1. do all this within the `dm` object by zooming to a table and manipulating it while maintaining the key relations whenever possible.
+
+We will explore the first approach here. For the second approach, see `vignette("tech-dm-zoom")`.
+
+## Enabling {dplyr}-workflow within a `dm`
+
+Subsetting a table of a `dm` object produces an object of class `dm_keyed_tbl`.
+
+{dm} provides methods for many of the {dplyr}-verbs for a `dm_keyed_tbl` which behave the way you are used to, affecting only the table and leaving the rest of the `dm` untouched.
+
+When you are finished with transforming the table, there are three options to proceed:
+
+1. use `dm()` if you want to replace the original tables with the new table
+1. use `new_dm()`/`dm()` if you want to create a brand new `dm` object 
+
+When employing these options, the resulting tables in the `dm` will have all the primary and foreign keys available that could be tracked from the original table.
+
+## Examples
+
+So much for the theory, but how does it look and feel?
+To explore this, we once more make use of our trusted {nycflights13} data.
+
+### Use case 1: Add a new column to an existing table
+
+Imagine you want to have a column in `flights`, specifying if a flight left before noon or after.
+Just like with {dplyr}, we can tackle this with `mutate()`.
+Let us do this step by step:
+
+```{r zoom}
+library(dm)
+library(dplyr)
+
+flights_dm <- dm_nycflights13()
+flights_dm
+
+flights_dm_tbl <-
+  flights_dm %>% 
+  dm_get_tables(keyed = TRUE)
+
+flights_tbl <- flights_dm_tbl$flights
+
+# The print output for a `dm_keyed_tbl` looks very much like that from a normal
+# `tibble`, with additional details about keys.
+flights_tbl
+
+flights_tbl_mutate <-
+  flights_tbl %>%
+  mutate(am_pm_dep = if_else(dep_time < 1200, "am", "pm")) %>%
+  # in order to see our changes in the output we use `select()` for reordering
+  # the columns
+  select(year:dep_time, am_pm_dep, everything())
+
+flights_tbl_mutate
+```
+
+To update the original `dm` with a new `flights` table we use `dm()`:
+
+```{r zoom2, error=TRUE}
+updated_flights_dm <- dm(
+  flights = flights_tbl_mutate,
+  flights_dm[c("airlines", "airports", "planes", "weather")]
```

Comment:
should this line use `dm_select_tbl()` to look tidy like the rest?

Reply:
same questions for all the code lines where this pattern appears.

Reply:
This should be `flights_dm_tbl`, at least currently.

Reply:
I mean ` flights_dm[c("airlines", "airports", "planes", "weather")]` could be `dm_select_tbl(dm, airlines, airports, planes, weather)`